### DISCUSSION TITLE
Pre-release v2.0.0

### DISCUSSION THREAD

**Discussion by kylechui** [kylechui/nvim-surround#141] [Announcements]
Hi everybody! This thread will be dedicated to discussing the upcoming `v2.0.0` and its new feature set/tweaked user configuration interface. Please feel free to express your thoughts on the update in the comments below: suggestions, bug fixes, and general comments are all welcome! I would really like to stabilize the user configuration interface once and for all, so if there are names that are even *slightly* confusing/seem out of place, please tag me and let me know.

Here's the list of [breaking changes](https://github.com/kylechui/nvim-surround/issues/77#issuecomment-1215932210) since `v1.0.0`. They're all relatively minor renames, so hopefully it's not too bad :)

## New Features

* Adds first-class Tree-sitter support for finding parent selections; see `:h nvim-surround.config.get_selection()`
* Filetype-specific surrounds (that utilize Tree-sitter) can be set by default
* `nvim-surround` indentation can be disabled by setting the new key `indent_lines = false`; see `:h nvim-surround.config.indent_lines`

Here's a small example of how the new Tree-sitter nodes can be used to discern what patterns could not:

https://user-images.githubusercontent.com/48545987/184789151-5c66b9c9-7c46-45a0-a0a6-e2a821addf35.mp4

I'm still not 100% sure on whether or not I want to set up filetype-specific surrounds by default, but the list that I was planning on using [can be found here](https://github.com/kylechui/nvim-surround/blob/main/lua/nvim-surround/filetype.lua).

**Comment by echasnovski** [comment]
As far as I can tell, the only purpose of filetype setup is to declare which tree-sitter node defines target object. Is this the case?

If it is indeed, I would recommend using more tree-sitter native approach to solving this problem by relying on queries. Here how this is done in [nvim-treesitter/nvim-treesitter-textobjects](https://github.com/nvim-treesitter/nvim-treesitter-textobjects/tree/master/queries). Basically, for every filetype there is a file which with help of tree-sitter query magic defines *captures* which can be used along with other nodes. You can either create your own queries (maybe by cherry-picking ones from nvim-treesitter-textobjects) with "namespaced" captures or tell users how to get/set those queries. This is the way I decided to go in 'mini.ai' and soon will be the case in 'mini.surround'.

With this approach there is no need to create all these autocommands but rather use something like `config.get_selection({ node = "@call.outer" })`.

**Reply by kylechui** [reply]
Thanks a bunch for pointing this out! I had no idea that queries could be used in this way. I'm still debating on whether or not I should include "more opinionated setup" into `filetype`, e.g. `c` for LaTeX "commands", but that remains to be seen (I don't really have a clean way of even handling all of user-defined global maps, user-defined buffer-local maps, default global maps, default buffer-local maps). It's all somewhat of a mess right now. I'm amazed that you even have the time to check out my plugin haha, thanks again for all the knowledge you've shared with me :sparkling_heart:

**Reply by echasnovski** [reply]
No problem. You asked for help on Reddit and I happen to have a recent relevant experience here :) 
For what it's worth, I really think that current approach with many filetype autocommands is suboptimal. Even if you decide to not use queries, having all this setup in 'after/ftplugin' directory looks like a cleaner approach to me.

LaTeX also has tree-sitter support and [some queries](https://github.com/nvim-treesitter/nvim-treesitter-textobjects/blob/master/queries/latex/textobjects.scm). Maybe this will be useful.

**Reply by kylechui** [reply]
I would be inclined to agree; my original reason for even doing this in the first place was to have things "just work" out-of-the-box, instead of requiring people to copy-paste things in that are:
1. Perhaps not easily understood, or
2. Very repetitive (as are the current function call nodes)

Perhaps by using queries the code duplication can be cut down even more, to the point where the configuration isn't too hard for a regular user to maintain.

I just want to say again that I really appreciate all the help/advice you've given me, in spite of the fact that we maintain "rival" plugins :laughing:

**Reply by clason** [reply]
LaTeX also has VimTeX with its own surround objects. Virtually everyone using LaTeX professionally has (or should have) VimTeX installed -- whose features are incompatible with tree-sitter. So I wouldn't spend much time on that, personally.

**Reply by kylechui** [reply]
That's also a good point; I found that out recently when somebody asked me to help with a command-style LaTeX surround that VimTeX already has `dsc/csc` commands built-in.

P.S. I recognize your name from Neovim core; thanks for helping to create such a great editor!

**Reply by kylechui** [reply]
@echasnovski I think I just got a rudimentary version using queries working; it seems better from the perspective that things "just work" for all filetypes, but it looks like I'll have to require `nvim-treesitter-textobjects` as a "dependency", unless I decide to copy over all of the queries and have them just built-in to `nvim-surround`. Am I misunderstanding anything? Thanks again!

**Reply by echasnovski** [reply]
> @echasnovski I think I just got a rudimentary version using queries working; it seems better from the perspective that things "just work" for all filetypes, but it looks like I'll have to require `nvim-treesitter-textobjects` as a "dependency", unless I decide to copy over all of the queries and have them just built-in to `nvim-surround`. Am I misunderstanding anything? Thanks again!

Yeah, that's right: in order to use queries, user should have them *somewhere*. In 'mini.nvim' I opted into [telling users how to set them up](https://github.com/echasnovski/mini.nvim/blob/475edc8c332c1a4df4b30fad6bf2db98136a0a59/lua/mini/ai.lua#L797-L806), but this is more in line with general design decision of not having anything language-specific inside of 'mini.nvim'.

Couple of notes:
- I am not sure what will happen if you copy those query files verbatim with the same captures. As user might have both 'nvim-treesitter-textobjects' and 'nvim-surround' installed, there might be conflicting captures and I don't know how it gets resolved. Besides, queries are constantly updating there (not a big deal for only `call.outer`, though).
- Queries from 'nvim-treesitter-textobjects' often rely on custom directives (special rules defining query logic) to make captures, so this also means a dependency on 'nvim-treesitter'. Doesn't seem to be a big deal in your case since you already using it for query logic.
- [This](https://github.com/kylechui/nvim-surround/blob/c057fb81f1496a88722e201eeb71bba06d532076/lua/nvim-surround/config.lua#L179-L181) looks very redundant. `pcall(require, 'nvim-treesitter')` should do the job.

**Reply by clason** [reply]
I would strongly suggest to depend on nvim-treesitter-textobjects. Queries can be highly specific to a certain parser _commit_, and these get updated constantly in nvim-treesitter. Every plugin that ships their own queries must make sure that they keep them up-to-date -- which is hard enough for nvim-treesitter-textobjects! Also, having two plugins ship the same (type of) query can lead to ugly problems when they are out of sync (or when trying to debug errors).

As long as its an optional dependency (meaning you only have to add them if you want the fancier surround functions), I see this as the best option.

**Reply by kylechui** [reply]
That sounds good to me; thanks again to both of you for your help! I'll keep tinkering with the configuration/setup options to see if I can find something that's a good mix of easy to use and extensible. I also like the idea of keeping nvim-treesitter-textobjects as an optional dependency that just unlocks "more" functionality. While I think this is a good new "search engine" for `find`ing "parent" selections for the surrounds, I'm still curious about whether queries can be used to actually "narrow down the search space", e.g. to find the left/right pair. As of right now, it looks like I have many ways of finding parent selections (Vim text-object, Lua pattern, Treesitter-node, and now query), but still am stuck on using Lua patterns for actually "filtering" the parent selection down to the left/right pair.

Edit: You two sharing your "expertise" on this with me (someone that knows very little) is a big reason why I love this community so much, and really motivates me :)

**Reply by clason** [reply]
Enthusiastic plugin devs such as you are the reason the community is great ;)

To be honest, I think tree-sitter queries are a bit of a blunt tool for finding "standard" left/right pairs -- nodes might be useful for that (depending on the parser) but that should be tested whether it's actually faster or more robust than just patterns.

I'd suggest looking at the current tree-sitter-textobjects and considering whether and how it makes sense to use these directly as surround objects -- for example, "csf" could `c`hange the `s`urrounding `f`unction, allowing you to rename it or something (whatever makes sense). Or maybe conditionals could be changed from `for` to `while` etc. (Basically, anything where changing the "top" requires a deterministic(!) change to the "bottom" -- like `for`-`endfor` to `while`-`endwhile`.)

**Reply by kylechui** [reply]
Ah yeah so `csf` is already "built-in" for renaming functions; the idea is that deletion/changing is a two-step process:
1. Find the "parent" selection *somehow*. In this particular case, we are using Tree-sitter queries, but this works in general with Lua patterns or built-in Vim text-objects as well.
2. "Filter" down the parent selection to a left/right pair by using a Lua pattern + match groups.

For example, what's already built-in is:
* Deleting a function call matches the "left" delimiter as everything up to *and including* the first opening parenthesis, and the "right" delimiter as just the closing parenthesis
* Changing a function call matches the "left" delimiter as everything up to *but not including* the first opening parenthesis, and nothing for the right delimiter

This two-step approach allows for "modularity" in the sense that if no Tree-sitter parser is detected for the current file (or something to that effect), then we can use Lua patterns (e.g. `%w+%b()` as a simple example) to try and find the nearest function call to serve as the parent selection. However if a parser *is* detected then we can use that, as it *should* provide better coverage of test cases like `smiley(":)")` being incorrectly picked up by pattern-matching.

While I've tried to cover as many edge cases as possible by utilizing match groups instead of assuming that the left/right pair *must* be at the beginning/end of the parent selection, I have the gut feeling that patterns aren't perfect. My comment earlier was just me thinking out loud about potential ways to improve no. 2 by somehow using queries to match "sub-patterns" in the parent query, as opposed to using Lua patterns to "filter" it down. In a sense just thinking of potential ways to perform step 2, just as there are now 3-4 ways to perform step 1.

**Reply by echasnovski** [reply]
One way to do this (which I am implementing in 'mini.surround' *literally* right now) is to have user supply not only `call.outer` but also a `call.inner` (basically, like in 'mini.ai' for `a` and `i` textobjects). Then surrounding is basically a difference between "around" and "inside" textobjects. This is not as straightforward as it should be (because of nvim-treesitter/nvim-treesitter-textobjects#195), but I'll make a PR tomorrow to resolve it.

**Reply by kylechui** [reply]
@echasnovski I've just added that functionality in, but with a twist: instead of just specifying a query for the "insides" of the surround, I've elected to make it a function that returns any arbitrary selection. The good thing about this is that the user can then re-use any of the methods used to find "parent selections" mentioned before to find the "inner" selection, and then use the difference as the delimiter pair. Another benefit of doing it this way as opposed to directly using queries is that the user can apply some "post-processing" to the delimiter pair, e.g. [here](https://github.com/kylechui/nvim-surround/blob/e2c22a62fe001eb7ef3bf088f4e0c439c9f9eefd/lua/nvim-surround/config.lua#L209) I "shift" the selections around in order to account for the fact that `@call.inner` is busted.

**Reply by echasnovski** [reply]
> e.g. [here](https://github.com/kylechui/nvim-surround/blob/e2c22a62fe001eb7ef3bf088f4e0c439c9f9eefd/lua/nvim-surround/config.lua#L209) I "shift" the selections around in order to account for the fact that @call.inner is busted.

Problem is that `@call.inner` might be "correct" and this will lead to wrong behavior. While there is no way for you as maintainer to know this beforehand. Maybe for some language it is already correct? Or as in Lua now it correctly catches the case `require"nvim-surround"`.
Dealing with language specific cases is very tricky.

**Reply by kylechui** [reply]
That's a very good point; I eagerly await your PR and will switch to the "correct" implementation once that's in place.

**Reply by kylechui** [reply]
Ok I've finally had some more time to think about how exactly I want to implement things; so far I have two ideas for how queries could be integrated:
1) Use existing queries. The user would have to specify which query they would like to call, as well as the source. E.g. for `nvim-treesitter-textobjects`, the user would need to specify `@call.outer` and `textobjects`
2) Have the user script new queries "on-the-fly". The user would provide the S-expression that represents the query, as well as which capture group from the query they would like to match for. For example, the query string could be
```scheme
(function_call
  (arguments
    (function_call))) @my_capture
```
and the capture group would be `my_capture`. The idea here is that even if you want to match for a large section, you could capture a subset to be the actual retrieved selection. It's analogous to using capture clauses for Lua pattern-matching in my current setup.

While I think case 2 can be done as a subset of case 1 (by simply asking the user to write their queries in a file, instead of directly in the configuration), I think it makes sense to allow them to put all configuration-related scripts in one spot. One of the problems I'm facing now is just what to actually *call* these things.

Right now I feel like since capture groups are provided for both, it might make sense to have only three keys, where the `capture` key is "shared" between two use cases. For example:
```lua
-- nvim-treesitter-textobjects, case #1
M.get_selection({ query = { capture = "call.outer", type = "textobjects" } })
-- on-the-fly, case #2
M.get_selection({ query = { capture = "my_capture", sexpr = "(function_call) @my_capture" } })
```

In the above I'm using the keys `capture`, `type`, and `sexpr` to get particular selections. As always, I would appreciate any feedback on the names!

Edit: This would allow for [selecting outer function calls](https://github.com/nvim-treesitter/nvim-treesitter-textobjects/issues/206), for example via the first query I provided above.

**Reply by echasnovski** [reply]
> While I think case 2 can be done as a subset of case 1 (by simply asking the user to write their queries in a file, instead of directly in the configuration), I think it makes sense to allow them to put all configuration-related scripts in one spot.

I am not sure how you plan this to work. Queries highly depend on language parser and need to be configured as such. There is a built-in mechanism for that (query files in 'queries' directory) which you use in "case 1". Do you want to go back to autocommands and filetypes but now make them optional?

**Reply by clason** [reply]
> There is a built-in mechanism for that (query files in 'queries' directory) 

I'm sorry, but no -- that just matches queries to parser _names_ and can't handle different parser _versions_ (which is a frequent pain point). We're still trying to figure out a solution to this problem.

**Reply by kylechui** [reply]
Ah wait, I just realized that queries are already filetype-specific like they're already in `ftplugin` :sweat_smile: Forget case 2 then; users can just provide their own queries by calling them from their own config files. Thanks for the sanity check haha

Edit: There could be some merit to defining a query for a specific buffer instead of an entire filetype, but to my knowledge the query doesn't really do anything unless it is "referenced"(?) so it should be fine with just case 1 here.

**Reply by echasnovski** [reply]
> I'm sorry, but no -- that just matches queries to parser names and can't handle different parser versions (which is a frequent pain point). We're still trying to figure out a solution to this problem.

Yes, that's true. What I really meant was more in terms of "configure query in `FileType` autocommand" vs "put query in 'after/queries/\<language>/custom.scm'".

**Comment by ggandor** [comment]
I think `_cur` and `_line` as suffixes are _really_ perplexing. "Cursor"? "Current..." what? `_cur`->`current_line` and `_line`->`_linewise` (or maybe `_around_line`?) would be much clearer.

**Reply by kylechui** [reply]
As per `:h nvim-surround.config.keymaps`:
```
The `keymaps` table defines the keymaps used to perform surround actions. The
general rule is that if the key ends in "_line", the delimiter pair is added
on new lines. If the key ends in "_cur", the surround is performed around the
current line.
```
I agree that the names could probably be a bit better; however I would like to have a comprehensive set of names for all normal mode binds (including the current `cur_line`) before I decide to make any changes.
